File: crates/static-analysis-kernel/src/analysis/ddsa_lib/js/ts_node.js
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License, Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.

import { op_ts_node_text } from "ext:core/ops";

/**
 * A non-zero integer assigned by the Rust static-analysis-kernel.
 *
 * This is not stable across analyses,
 * @typedef {number} NodeId
 */

/**
 * The tree-sitter u16 `TSSymbol`, specific to a tree-sitter Language.
 * @typedef {number} NodeTypeId
 */

/**
 * A specific point (1-based line number, 1-based column number) within a source text.
 * @typedef {Object} Position
 * @property {number} line
 * @property {number} col
 */

/**
 * An object representing a node within a `tree-sitter Tree, as well as functions to access metadata about
 * both itself and its relationship to its context.
 */
export class TreeSitterNode {
    /**
     * @param {NodeId} id
     * @param {number} startLine
     * @param {number} startCol
     * @param {number} endLine
     * @param {number} endCol
     * @param {NodeTypeId} nodeTypeId
     */
    constructor(id, startLine, startCol, endLine, endCol, nodeTypeId) {
        /**
         * The id of this node, assigned by the Rust static analysis kernel.
         * @type {NodeId}
         * @private
         * */
        this.id = id;
        /**
         * @type {number}
         * @private
         */
        this._startLine = startLine;
        /**
         * @type {number}
         * @private
         */
        this._startCol = startCol;
        /**
         * @type {number}
         * @private
         */
        this._endLine = endLine;
        /**
         * @type {number}
         * @private
         */
        this._endCol = endCol;
        /** @type {NodeTypeId} */
        this._typeId = nodeTypeId;
        /**
         * A lazily-allocated start {@link Position}, created and/or returned when requested via the {@link TreeSitterNode.start} getter.
         * @type {Position | undefined}
         * @private
         */
        this._cachedStart = undefined;
        /**
         * A lazily-allocated end {@link Position}, created and/or returned when requested via the {@link TreeSitterNode.end} getter.
         * @type {Position | undefined}
         * @private
         */
        this._cachedEnd = undefined;
        /**
         * A lazily-allocated string of the text that this node spans. This will only be stored if the text
         * is requested via the {@link TreeSitterNode.text} getter.
         * @type {string | undefined}
         * @private
         */
        this.__js_cachedText = undefined;
    }

    /**
     * A getter to return the text that this `TreeSitterNode` spans.
     * @returns {string}
     *
     * @remarks
     * This lazily makes a call to Rust to retrieve the node's text. Subsequent calls to this getter will
     * return the cached value.
     */
    get text() {
        if (this.__js_cachedText === undefined) {
            const opResult = op_ts_node_text(this.id);
            if (opResult === null) {
                // This branch is only accessible if this node's `id` is mutated such that a non-existent id
                // is passed into the op. In this case, return undefined.
                return undefined;
            } else {
                this.__js_cachedText = opResult;
            }
        }
        return this.__js_cachedText;
    }

    /**
     * A getter to return the start {@link Position} of this node.
     * Note that this getter returns a cached object -- the caller should not mutate it.
     * @returns {Position}
     */
    get start() {
        if (this._cachedStart === undefined) {
            this._cachedStart = buildPosition(this._startLine, this._startCol);
        }
        return this._cachedStart;
    }

    /**
     * A getter to return the end {@link Position} of this node.
     * Note that this getter returns a cached object -- the caller should not mutate it.
     * @returns {Position}
     */
    get end() {
        if (this._cachedEnd === undefined) {
            this._cachedEnd = buildPosition(this._endLine, this._endCol);
        }
        return this._cachedEnd;
    }

    /**
     * A getter to return the string version of this node's type.
     * @returns {string}
     *
     * @example
     * ```javascript
     * "function_declaration"
     * ```
     */
    get type() {
        // Note: the map lookup should only return undefined if either `this._typeId` or the symbol map were mutated.
        // Although that should never happen, we handle it by returning an empty string.
        return globalThis.__RUST_BRIDGE__context.tsLangCtx.nodeType.get(this._typeId) ?? "";
    }

    /**
     * A getter to return the named children of this tree-sitter node.
     * NOTE: This is deprecated, because it is a compatibility layer to support the stella API.
     * Do not rely on this, as it will be removed.
     *
     * @returns {Array<TreeSitterNode | TreeSitterFieldChildNode>}
     * @deprecated
     */
    get children() {
        return globalThis.ddsa.getChildren(this);
    }

    /**
     * A getter to return the string version of this node's type.
     * NOTE: This is deprecated, because it is a compatibility layer to support the stella API.
     * Do not rely on this, as it will be removed.
     *
     * @returns {string}
     * @deprecated
     */
    get astType() {
        return this.type;
    }
}

/**
 * The non-zero tree-sitter u16 field id for a child node in relation to a parent.
 * @typedef {number} FieldId
 */

/**
 * An object with the same interface as a {@link TreeSitterNode} that is the child of another node, and
 * additionally is tagged with a field name.
 * It does not mutate the `TreeSitterNode` that it wraps.
 *
 * @extends TreeSitterNode
 */
export class TreeSitterFieldChildNode {
    static [Symbol.hasInstance](instance) {
        // Because we set the prototype to the wrapped object, we lose the ability to call instanceof on `TreeSitterFieldChildNode`.
        // To implement it, we use the fact `TreeSitterFieldChildNode` uniquely has `_fieldId`.
        return instance instanceof TreeSitterNode && instance._fieldId !== undefined;
    }

    /**
     * @param {TreeSitterNode} tsNode The `TreeSitterNode` that will be wrapped.
     * @param {number} fieldId The field id of this child node.
     */
    constructor(tsNode, fieldId) {
        /**
         * The field id of this child node.
         * @type {FieldId}
         * @private
         * */
        this._fieldId = fieldId;

        /**
         * The field name associated with this child.
         * @type {string}
         * @readonly
         */
        // NOTE: We need to use `Object.defineProperty` so we can assign the property to this specific instance.
        // We can't use the `get` keyword, like:
        //
        // class TreeSitterFieldChildNode {
        //     get fieldName() { /* ... */ }
        // }
        //
        // because `fieldName` would then be assigned to the prototype, not the instance. That wouldn't work for us because
        // at the end of this constructor, we assign the prototype to the wrapped `TreeSitterNode`.
        Object.defineProperty(this, "fieldName", {
            get() {
                // Note: This will only return `undefined` if the `_fieldId` field is mutated. Because this is an unsupported
                // edge case, it's excluded from the type signature for clarity.
                return globalThis.__RUST_BRIDGE__context.tsLangCtx.field.get(this._fieldId);
            },
            enumerable: false,
            configurable: false,
        });

        // (Note: The use of `?? {}` is only because the Rust unit tests inspect the instance by invoking the constructor with no params)
        Object.setPrototypeOf(this, tsNode ?? {});
    }
}

/**
 * Creates a new {@link Position} from a line and row number.
 * @param {number} lineNumber The 1-based line number
 * @param {number} columnNumber The 1-based column number
 * @returns {Position}
 */
function buildPosition(lineNumber, columnNumber) {
    return {
        line: lineNumber,
        col: columnNumber
    };
}

File: crates/static-analysis-kernel/src/analysis/ddsa_lib/js/ts_node.rs
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License, Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.

use crate::analysis::ddsa_lib::common::{
    load_function, v8_uint, Class, DDSAJsRuntimeError, Instance, NodeId,
};
use deno_core::v8;
use deno_core::v8::HandleScope;
use std::marker::PhantomData;

/// A deserialized JavaScript object representation of a [`tree_sitter::Node`].
#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
pub struct TreeSitterNode<T> {
    id: u32,
    start_line: u32,
    start_col: u32,
    end_line: u32,
    end_col: u32,
    /// The type of node this is. This corresponds to [`tree_sitter::Node::kind_id`].
    node_type_id: u16,
    _pd: PhantomData<T>,
}

impl<T> TreeSitterNode<T> {
    /// Converts the provided [`tree_sitter::Node`] into a `TreeSitterNode`, assigning the provided id.
    pub fn from_ts_node(id: NodeId, node: tree_sitter::Node) -> Self {
        // NOTE: We normalize the 0-based `tree_sitter::Point` to be 1-based.
        fn normalize_ts_point_num(num: usize) -> u32 {
            num as u32 + 1
        }
        let tree_sitter::Point {
            row: start_line,
            column: start_col,
        } = node.start_position();
        let tree_sitter::Point {
            row: end_line,
            column: end_col,
        } = node.end_position();
        Self {
            id,
            start_line: normalize_ts_point_num(start_line),
            start_col: normalize_ts_point_num(start_col),
            end_line: normalize_ts_point_num(end_line),
            end_col: normalize_ts_point_num(end_col),
            node_type_id: node.kind_id(),
            _pd: PhantomData,
        }
    }
}

/// A function representing the ES6 class `TreeSitterNode`.
#[derive(Debug)]
pub struct TreeSitterNodeFn<T>(v8::Global<v8::Function>, PhantomData<T>);

impl TreeSitterNodeFn<Class> {
    pub const CLASS_NAME: &'static str = "TreeSitterNode";

    /// Creates a new [`v8::Global`] function by loading [`Self::CLASS_NAME`] from the `scope`.
    pub fn try_new(scope: &mut HandleScope) -> Result<Self, DDSAJsRuntimeError> {
        load_function(scope, Self::CLASS_NAME).map(|func| Self(func, PhantomData))
    }

    /// Creates a new instance of the `TreeSitterNode` class.
    pub fn new_instance<'s>(
        &self,
        scope: &mut HandleScope<'s>,
        ts_node: TreeSitterNode<Instance>,
    ) -> v8::Local<'s, v8::Object> {
        let id = v8_uint(scope, ts_node.id).into();
        let start_line = v8_uint(scope, ts_node.start_line).into();
        let start_col = v8_uint(scope, ts_node.start_col).into();
        let end_line = v8_uint(scope, ts_node.end_line).into();
        let end_col = v8_uint(scope, ts_node.end_col).into();
        let node_type_id = v8_uint(scope, ts_node.node_type_id as u32).into();
        let args = [id, start_line, start_col, end_line, end_col, node_type_id];
        self.0
            .open(scope)
            .new_instance(scope, &args[..])
            .expect("class constructor should not throw")
    }
}

/// The name of the ES6 class that wraps a `TreeSitterNode`, providing additional metadata about
/// the node as a child of another node.
///
/// This is only constructed in JavaScript, so we only track the name here.
const FIELD_CHILD_NODE_CLASS_NAME: &str = "TreeSitterFieldChildNode";

#[cfg(test)]
mod tests {
    use crate::analysis::ddsa_lib::common::{v8_interned, v8_string, Class, Instance};
    use crate::analysis::ddsa_lib::context::ts_lang;
    use crate::analysis::ddsa_lib::js::ts_node::FIELD_CHILD_NODE_CLASS_NAME;
    use crate::analysis::ddsa_lib::js::{TreeSitterNode, TreeSitterNodeFn};
    use crate::analysis::ddsa_lib::test_utils::{
        attach_as_global, cfg_test_runtime, js_class_eq, js_instance_eq, make_stub_root_context,
        try_execute,
    };
    use crate::analysis::tree_sitter::get_tree_sitter_language;
    use crate::model::common::Language;
    use deno_core::v8;
    use std::marker::PhantomData;

    #[test]
    fn js_properties_canary() {
        let expected = &[
            // Variables
            "id",
            "_startLine",
            "_startCol",
            "_endLine",
            "_endCol",
            "_typeId",
            "_cachedStart",
            "_cachedEnd",
            "__js_cachedText",
            // Methods
            "text",
            "type",
            "start",
            "end",
            "children",
        ];
        assert!(js_instance_eq(TreeSitterNodeFn::CLASS_NAME, expected));
        let expected = &[];
        assert!(js_class_eq(TreeSitterNodeFn::CLASS_NAME, expected));

        let expected = &[
            // Variables
            "_fieldId",
            // Methods
            "fieldName",
        ];
        assert!(js_instance_eq(FIELD_CHILD_NODE_CLASS_NAME, expected));
        let expected = &[];
        assert!(js_class_eq(FIELD_CHILD_NODE_CLASS_NAME, expected));
    }

    /// Tests that the line and column number of the node is 1-based. This test is necessary because
    /// [`tree_sitter::Point`] is 0-based, whereas we use a 1-based number.
    #[test]
    fn ts_node_line_col_one_based() {
        let lang = get_tree_sitter_language(&Language::JavaScript);
        let mut parser = tree_sitter::Parser::new();
        parser.set_language(&lang).unwrap();

        let tree = parser.parse(r#"foo(bar, baz);"#, None).unwrap();
        let root = tree.root_node();

        assert_eq!(root.start_position().row, 0);
        assert_eq!(root.start_position().column, 0);
        assert_eq!(root.end_position().row, 0);
        assert_eq!(root.end_position().column, 14);

        let tree_sitter_node = TreeSitterNode::<Instance>::from_ts_node(0, root);

        assert_eq!(tree_sitter_node.start_line, 1);
        assert_eq!(tree_sitter_node.start_col, 1);
        assert_eq!(tree_sitter_node.end_line, 1);
        assert_eq!(tree_sitter_node.end_col, 15);
    }

    /// Tests that we use the [`tree_sitter::Node::kind_id`] instead of the [`tree_sitter::Node::grammar_id`], which
    /// has subtle differences.
    #[rustfmt::skip]
    #[test]
    fn ts_node_kind_id() {
        let lang = get_tree_sitter_language(&Language::JavaScript);
        let mut parser = tree_sitter::Parser::new();
        parser.set_language(&lang).unwrap();

        let text = r#"{ keyName: "value" }"#;
        let tree = parser.parse(text, None).unwrap();

        let query = tree_sitter::Query::new(&lang, "(pair key: (_) @cap)").unwrap();
        let mut cursor = tree_sitter::QueryCursor::new();
        let ts_node = cursor.matches(&query, tree.root_node(), text.as_bytes()).next().unwrap().captures.first().unwrap().node;
        // First make sure this test is properly constructed (an upstream tree_sitter grammar change
        // may invalidate this, so we need to confirm).
        // This should resolve to `assert_ne!("identifier", "property_identifier")`
        assert_ne!(ts_node.grammar_name(), ts_node.kind());

        let tree_sitter_node = TreeSitterNode::<Instance>::from_ts_node(0, ts_node);
        assert_eq!(tree_sitter_node.node_type_id, ts_node.kind_id());
    }

    /// Tests that the getter for `start` and `end` is lazily instantiated and returned.
    #[test]
    fn position_getters() {
        let mut runtime = cfg_test_runtime();
        let scope = &mut runtime.handle_scope();
        let js_class = TreeSitterNodeFn::<Class>::try_new(scope).unwrap();

        let base_ts_node = TreeSitterNode::<Instance> {
            id: 3,
            start_line: 123,
            start_col: 1,
            end_line: 456,
            end_col: 32,
            node_type_id: 8,
            _pd: PhantomData,
        };
        let v8_ts_node = js_class.new_instance(scope, base_ts_node);
        attach_as_global(scope, v8_ts_node, "TS_NODE");

        for (getter, cache_name, (expected_line, expected_col)) in [
            ("start", "_cachedStart", (123, 1)),
            ("end", "_cachedEnd", (456, 32)),
        ] {
            // Verify the cache behavior
            let s_cache_name = v8_interned(scope, cache_name);
            let v8_cached_pos = v8_ts_node.get(scope, s_cache_name.into()).unwrap();
            assert!(v8_cached_pos.is_undefined());

            // Return the value from the provided getter: e.g. `TS_NODE.start`.
            let code = format!("TS_NODE.{};", getter);
            let res = try_execute(scope, &code).unwrap();

            // Ensure we get a position object back
            let v8_returned_pos: v8::Local<v8::Object> = res.try_into().unwrap();
            let own_props = v8_returned_pos
                .get_own_property_names(scope, Default::default())
                .unwrap();
            assert_eq!(own_props.length(), 2);

            let s_line = v8_interned(scope, "line");
            let s_col = v8_interned(scope, "col");
            let v8_line = v8_returned_pos.get(scope, s_line.into()).unwrap();
            let v8_col = v8_returned_pos.get(scope, s_col.into()).unwrap();
            assert_eq!(v8_line.uint32_value(scope).unwrap(), expected_line);
            assert_eq!(v8_col.uint32_value(scope).unwrap(), expected_col);

            // Assert that node cached this position (i.e. the property is no longer undefined).
            let v8_cached_pos = v8_ts_node.get(scope, s_cache_name.into()).unwrap();
            assert!(v8_cached_pos.is_object());

            // And then assert that beyond just property value equality, the getter returns the
            // exact same object as the cached object.
            assert_eq!(v8_cached_pos.get_hash(), v8_returned_pos.get_hash())
        }
    }

    /// Tests that the `type` getter returns the node type as a name string.
    #[test]
    fn type_getter() {
        let mut runtime = cfg_test_runtime();
        let scope = &mut runtime.handle_scope();
        let js_class = TreeSitterNodeFn::<Class>::try_new(scope).unwrap();

        let lang_js = get_tree_sitter_language(&Language::JavaScript);
        // (Included to alert if tree_sitter grammar changes the node kind ids)
        const EXPECTED: (u16, &str) = (149, "variable_declarator");
        assert_eq!(lang_js.node_kind_for_id(EXPECTED.0), Some(EXPECTED.1));

        let stub_root_ctx = make_stub_root_context(scope, &[], "", "", Some(&lang_js));
        attach_as_global(scope, stub_root_ctx, "__RUST_BRIDGE__context");

        let base_ts_node = TreeSitterNode::<Instance> {
            id: 2,
            start_line: 456,
            start_col: 5,
            end_line: 456,
            end_col: 32,
            node_type_id: EXPECTED.0,
            _pd: PhantomData,
        };
        let v8_ts_node = js_class.new_instance(scope, base_ts_node);
        attach_as_global(scope, v8_ts_node, "TS_NODE");

        let code = "TS_NODE.type;";
        let ret_value = try_execute(scope, code).unwrap();
        assert_eq!(ret_value.to_rust_string_lossy(scope), EXPECTED.1);

        // And if the TSNode is mutated to have an invalid _typeId, it should return an empty string.
        let code = "TS_NODE._typeId = 99999; TS_NODE.type;";
        let ret_value = try_execute(scope, code).unwrap();
        assert_eq!(ret_value.to_rust_string_lossy(scope), "");
    }

    /// Tests that a [`FIELD_CHILD_NODE_CLASS_NAME`] wraps a [`TreeSitterNodeFn::CLASS_NAME`] with additional metadata,
    /// but that it provides all the same properties as the node itself.
    #[test]
    fn field_child_interface() {
        let mut runtime = cfg_test_runtime();
        let scope = &mut runtime.handle_scope();
        let js_class = TreeSitterNodeFn::<Class>::try_new(scope).unwrap();

        let lang_js = get_tree_sitter_language(&Language::JavaScript);
        // (Assertion included to alert if upstream tree-sitter grammar unexpectedly alters metadata)
        assert_eq!(lang_js.field_name_for_id(30), Some("optional_chain"));

        let base_ts_node = TreeSitterNode::<Instance> {
            id: 2,
            start_line: 1,
            start_col: 5,
            end_line: 1,
            end_col: 12,
            node_type_id: 0, // Unused
            _pd: PhantomData,
        };
        let v8_ts_node = js_class.new_instance(scope, base_ts_node);
        attach_as_global(scope, v8_ts_node, "TS_NODE");

        // We test the prototype, and then ensure properties and getters have the correct "this" context.
        let code = r#"
const assert = (val, msg) => { if (!val) throw new Error(msg); };

const child = new TreeSitterFieldChildNode(TS_NODE, 30);
assert(child instanceof TreeSitterNode, "should be a TreeSitterNode instance");
assert(child instanceof TreeSitterFieldChildNode, "should (also) be a TreeSitterFieldChildNode");
assert(child._startLine === 1, "_startLine value incorrect");
assert(child._startCol === 5, "_startCol value incorrect");
assert(child.start.col === 5, "start getter value incorrect");
assert(child._fieldId === 30, "_fieldId value incorrect");
123;
"#;
        let res = try_execute(scope, code).unwrap();
        assert!(res.is_uint32() && res.uint32_value(scope).unwrap() == 123);
    }
}
